File: frontend/lib/api/resource-sets.ts
import api from "@/lib/api";
import type {
  ResourceSetCreate,
  ResourceSetListItem,
  ResourceSetResponse,
  ResourceSetUpdate,
} from "@/types/resource-set";

export const getResourceSets = async (): Promise<ResourceSetListItem[]> => {
  const { data } = await api.get("/api/resource-sets/");
  return data;
};

export const createResourceSet = async (
  payload: ResourceSetCreate,
): Promise<ResourceSetResponse> => {
  const { data } = await api.post("/api/resource-sets/", payload);
  return data;
};

export const getResourceSet = async (
  id: string,
): Promise<ResourceSetResponse> => {
  const { data } = await api.get(`/api/resource-sets/${id}`);
  return data;
};

export const updateResourceSet = async (
  id: string,
  payload: ResourceSetUpdate,
): Promise<ResourceSetResponse> => {
  const { data } = await api.put(`/api/resource-sets/${id}`, payload);
  return data;
};

export const deleteResourceSet = async (id: string): Promise<void> => {
  await api.delete(`/api/resource-sets/${id}`);
};

export const publishResourceSet = async (
  id: string,
): Promise<ResourceSetResponse> => {
  const { data } = await api.post(`/api/resource-sets/${id}/publish`);
  return data;
};

export const archiveResourceSet = async (
  id: string,
): Promise<ResourceSetResponse> => {
  const { data } = await api.post(`/api/resource-sets/${id}/archive`);
  return data;
};

File: backend/app/routes/resource_sets.py
import uuid
from typing import List

from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models.user import User
from app.schemas.resource_set import (
    ResourceSetCreate,
    ResourceSetListItem,
    ResourceSetResponse,
    ResourceSetUpdate,
)
from app.services.resource_set_service import ResourceSetService
from app.utils.dependencies import get_current_teacher, get_language

router = APIRouter(prefix="/api/resource-sets", tags=["Resource Sets"])


@router.get("/", response_model=List[ResourceSetListItem])
async def list_resource_sets(
    db: AsyncSession = Depends(get_db),
    teacher: User = Depends(get_current_teacher),
    language: str = Depends(get_language),
):
    return await ResourceSetService.list_resource_sets(db, teacher.id, language)


@router.post(
    "/", response_model=ResourceSetResponse, status_code=status.HTTP_201_CREATED
)
async def create_resource_set(
    data: ResourceSetCreate,
    db: AsyncSession = Depends(get_db),
    teacher: User = Depends(get_current_teacher),
):
    return await ResourceSetService.create_resource_set(db, teacher.id, data)


@router.get("/{resource_set_id}", response_model=ResourceSetResponse)
async def get_resource_set(
    resource_set_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    teacher: User = Depends(get_current_teacher),
):
    return await ResourceSetService.get_resource_set(db, resource_set_id, teacher.id)


@router.put("/{resource_set_id}", response_model=ResourceSetResponse)
async def update_resource_set(
    resource_set_id: uuid.UUID,
    data: ResourceSetUpdate,
    db: AsyncSession = Depends(get_db),
    teacher: User = Depends(get_current_teacher),
):
    return await ResourceSetService.update_resource_set(
        db, resource_set_id, teacher.id, data
    )


@router.delete("/{resource_set_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_resource_set(
    resource_set_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    teacher: User = Depends(get_current_teacher),
):
    await ResourceSetService.delete_resource_set(db, resource_set_id, teacher.id)


@router.post("/{resource_set_id}/publish", response_model=ResourceSetResponse)
async def publish_resource_set(
    resource_set_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    teacher: User = Depends(get_current_teacher),
):
    return await ResourceSetService.publish_resource_set(
        db, resource_set_id, teacher.id
    )


@router.post("/{resource_set_id}/archive", response_model=ResourceSetResponse)
async def archive_resource_set(
    resource_set_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    teacher: User = Depends(get_current_teacher),
):
    return await ResourceSetService.archive_resource_set(
        db, resource_set_id, teacher.id
    )
